[Build] error: array index 7 is past the end of the array (that has type '__m256[4]')

### Describe the issue

Build fails with version 1.20.1 while it succeeded with version 1.18.2:

```
In file included from /usr/ports/misc/onnxruntime/work/onnxruntime-1.20.1/onnxruntime/core/mlas/lib/sqnbitgemm_kernel_avx512.cpp:25:
In file included from /usr/ports/misc/onnxruntime/work/onnxruntime-1.20.1/onnxruntime/core/mlas/lib/sqnbitgemm_kernel_avx512_int8_blklen16.h:9:
In file included from /usr/ports/misc/onnxruntime/work/onnxruntime-1.20.1/onnxruntime/core/mlas/lib/sqnbitgemm_kernel_avx512_int8_blklen32.h:8:
/usr/ports/misc/onnxruntime/work/onnxruntime-1.20.1/onnxruntime/core/mlas/lib/sqnbitgemm_kernel_avx2_int8_blklen32.h:954:49: error: array index 4 is past the end of the array (that has type '__m256[4]') [-Werror,-Warray-bounds]
  954 |                 __m128 acc_1 = FoldAccumulators(acc[4], acc[5], acc[6], acc[7]);
      |                                                 ^   ~
/usr/ports/misc/onnxruntime/work/onnxruntime-1.20.1/onnxruntime/core/mlas/lib/sqnbitgemm_kernel_avx2_int8_blklen32.h:822:13: note: array 'acc' declared here
  822 |             __m256 acc[NCols4];
      |             ^
/usr/ports/misc/onnxruntime/work/onnxruntime-1.20.1/onnxruntime/core/mlas/lib/sqnbitgemm_kernel_avx2_int8_blklen32.h:954:57: error: array index 5 is past the end of the array (that has type '__m256[4]') [-Werror,-Warray-bounds]
  954 |                 __m128 acc_1 = FoldAccumulators(acc[4], acc[5], acc[6], acc[7]);
      |                                                         ^   ~
/usr/ports/misc/onnxruntime/work/onnxruntime-1.20.1/onnxruntime/core/mlas/lib/sqnbitgemm_kernel_avx2_int8_blklen32.h:822:13: note: array 'acc' declared here
  822 |             __m256 acc[NCols4];
      |             ^
/usr/ports/misc/onnxruntime/work/onnxruntime-1.20.1/onnxruntime/core/mlas/lib/sqnbitgemm_kernel_avx2_int8_blklen32.h:954:65: error: array index 6 is past the end of the array (that has type '__m256[4]') [-Werror,-Warray-bounds]
  954 |                 __m128 acc_1 = FoldAccumulators(acc[4], acc[5], acc[6], acc[7]);
      |                                                                 ^   ~
/usr/ports/misc/onnxruntime/work/onnxruntime-1.20.1/onnxruntime/core/mlas/lib/sqnbitgemm_kernel_avx2_int8_blklen32.h:822:13: note: array 'acc' declared here
  822 |             __m256 acc[NCols4];
      |             ^
/usr/ports/misc/onnxruntime/work/onnxruntime-1.20.1/onnxruntime/core/mlas/lib/sqnbitgemm_kernel_avx2_int8_blklen32.h:954:73: error: array index 7 is past the end of the array (that has type '__m256[4]') [-Werror,-Warray-bounds]
  954 |                 __m128 acc_1 = FoldAccumulators(acc[4], acc[5], acc[6], acc[7]);
      |                                                                         ^   ~
/usr/ports/misc/onnxruntime/work/onnxruntime-1.20.1/onnxruntime/core/mlas/lib/sqnbitgemm_kernel_avx2_int8_blklen32.h:822:13: note: array 'acc' declared here
  822 |             __m256 acc[NCols4];
      |             ^
4 errors generated.
```

### Urgency

_No response_

### Target platform

FreeBSD 14.2

### Build script

cmake && gmake

### Error / output

Build fails:
```
In file included from /usr/ports/misc/onnxruntime/work/onnxruntime-1.20.1/onnxruntime/core/mlas/lib/sqnbitgemm_kernel_avx512.cpp:25:
In file included from /usr/ports/misc/onnxruntime/work/onnxruntime-1.20.1/onnxruntime/core/mlas/lib/sqnbitgemm_kernel_avx512_int8_blklen16.h:9:
In file included from /usr/ports/misc/onnxruntime/work/onnxruntime-1.20.1/onnxruntime/core/mlas/lib/sqnbitgemm_kernel_avx512_int8_blklen32.h:8:
/usr/ports/misc/onnxruntime/work/onnxruntime-1.20.1/onnxruntime/core/mlas/lib/sqnbitgemm_kernel_avx2_int8_blklen32.h:954:49: error: array index 4 is past the end of the array (that has type '__m256[4]') [-Werror,-Warray-bounds]
  954 |                 __m128 acc_1 = FoldAccumulators(acc[4], acc[5], acc[6], acc[7]);
      |                                                 ^   ~
/usr/ports/misc/onnxruntime/work/onnxruntime-1.20.1/onnxruntime/core/mlas/lib/sqnbitgemm_kernel_avx2_int8_blklen32.h:822:13: note: array 'acc' declared here
  822 |             __m256 acc[NCols4];
      |             ^
/usr/ports/misc/onnxruntime/work/onnxruntime-1.20.1/onnxruntime/core/mlas/lib/sqnbitgemm_kernel_avx2_int8_blklen32.h:954:57: error: array index 5 is past the end of the array (that has type '__m256[4]') [-Werror,-Warray-bounds]
  954 |                 __m128 acc_1 = FoldAccumulators(acc[4], acc[5], acc[6], acc[7]);
      |                                                         ^   ~
/usr/ports/misc/onnxruntime/work/onnxruntime-1.20.1/onnxruntime/core/mlas/lib/sqnbitgemm_kernel_avx2_int8_blklen32.h:822:13: note: array 'acc' declared here
  822 |             __m256 acc[NCols4];
      |             ^
/usr/ports/misc/onnxruntime/work/onnxruntime-1.20.1/onnxruntime/core/mlas/lib/sqnbitgemm_kernel_avx2_int8_blklen32.h:954:65: error: array index 6 is past the end of the array (that has type '__m256[4]') [-Werror,-Warray-bounds]
  954 |                 __m128 acc_1 = FoldAccumulators(acc[4], acc[5], acc[6], acc[7]);
      |                                                                 ^   ~
/usr/ports/misc/onnxruntime/work/onnxruntime-1.20.1/onnxruntime/core/mlas/lib/sqnbitgemm_kernel_avx2_int8_blklen32.h:822:13: note: array 'acc' declared here
  822 |             __m256 acc[NCols4];
      |             ^
/usr/ports/misc/onnxruntime/work/onnxruntime-1.20.1/onnxruntime/core/mlas/lib/sqnbitgemm_kernel_avx2_int8_blklen32.h:954:73: error: array index 7 is past the end of the array (that has type '__m256[4]') [-Werror,-Warray-bounds]
  954 |                 __m128 acc_1 = FoldAccumulators(acc[4], acc[5], acc[6], acc[7]);
      |                                                                         ^   ~
/usr/ports/misc/onnxruntime/work/onnxruntime-1.20.1/onnxruntime/core/mlas/lib/sqnbitgemm_kernel_avx2_int8_blklen32.h:822:13: note: array 'acc' declared here
  822 |             __m256 acc[NCols4];
      |             ^
4 errors generated.
```

### Visual Studio Version

_No response_

### GCC / Compiler Version

clang-18

--- a/cmake/CMakeLists.txt
+++ b/cmake/CMakeLists.txt
@@ -558,6 +558,7 @@ else()
   check_cxx_compiler_flag(-Wambiguous-reversed-operator HAS_AMBIGUOUS_REVERSED_OPERATOR)
   # -Winterference-size was added in GCC 13
   check_cxx_compiler_flag(-Winterference-size HAS_INTERFERENCE_SIZE)
+  check_cxx_compiler_flag(-Warray-bounds HAS_ARRAY_BOUNDS)
   check_cxx_compiler_flag(-Wbitwise-instead-of-logical HAS_BITWISE_INSTEAD_OF_LOGICAL)
   check_cxx_compiler_flag(-Wcast-function-type HAS_CAST_FUNCTION_TYPE)
   check_cxx_compiler_flag(-Wcatch-value HAS_CATCH_VALUE)

--- a/cmake/onnxruntime_config.h.in
+++ b/cmake/onnxruntime_config.h.in
@@ -3,6 +3,7 @@
 
 #pragma once
 
+#cmakedefine HAS_ARRAY_BOUNDS
 #cmakedefine HAS_BITWISE_INSTEAD_OF_LOGICAL
 #cmakedefine HAS_CAST_FUNCTION_TYPE
 #cmakedefine HAS_CATCH_VALUE

--- a/onnxruntime/core/mlas/lib/sqnbitgemm_kernel_avx2_int8_blklen32.h
+++ b/onnxruntime/core/mlas/lib/sqnbitgemm_kernel_avx2_int8_blklen32.h
@@ -1660,7 +1660,24 @@ MlasQ4Int8TileGemmKernelBlkLen32Avx2(
 
             if constexpr (NCols4 == 8) {
                 __m128 acc_0 = FoldAccumulators(acc[0], acc[1], acc[2], acc[3]);
+
+                // Clang is not happy with the code here, even if constexpr `NCols4 == 8` is always false in this context:
+                //
+                // In file included from .../onnxruntime/core/mlas/lib/sqnbitgemm_kernel_avx2.cpp:26:
+                // .../onnxruntime/core/mlas/lib/sqnbitgemm_kernel_avx2_int8_blklen32.h:1663:49: error: array index 4 is past the end of the array (that has type '__m256[4]') [-Werror,-Warray-bounds]
+                //  1663 |                 __m128 acc_1 = FoldAccumulators(acc[4], acc[5], acc[6], acc[7]);
+                //       |                                                 ^   ~
+                // .../onnxruntime/core/mlas/lib/sqnbitgemm_kernel_avx2_int8_blklen32.h:1531:13: note: array 'acc' declared here
+                //  1531 |             __m256 acc[NCols4];
+                //       |             ^
+#if defined(__clang__) && defined(HAS_ARRAY_BOUNDS)
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Warray-bounds"
+#endif
                 __m128 acc_1 = FoldAccumulators(acc[4], acc[5], acc[6], acc[7]);
+#if defined(__clang__) && defined(HAS_ARRAY_BOUNDS)
+#pragma clang diagnostic pop
+#endif
                 if (BiasPtr != nullptr) {
                     acc_0 = _mm_add_ps(acc_0, _mm_loadu_ps(BiasPtr));
                     acc_1 = _mm_add_ps(acc_1, _mm_loadu_ps(BiasPtr + 4));

